File: ChatGPT/Intermediárias/longest_palindromic_substring/longest_palindromic_substring.java
public class Main {
    // Método para expandir ao redor do centro e encontrar o palíndromo
    private static String expandAroundCenter(String s, int left, int right) {
        while (left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)) {
            left--;
            right++;
        }
        return s.substring(left + 1, right);
    }

    // Método para encontrar a maior substring palindrômica
    public static String longestPalindromicSubstring(String s) {
        if (s == null || s.length() < 1) {
            return "";
        }

        String longest = "";
        for (int i = 0; i < s.length(); i++) {
            // Palíndromo de comprimento ímpar
            String oddPalindrome = expandAroundCenter(s, i, i);
            if (oddPalindrome.length() > longest.length()) {
                longest = oddPalindrome;
            }
            // Palíndromo de comprimento par
            String evenPalindrome = expandAroundCenter(s, i, i + 1);
            if (evenPalindrome.length() > longest.length()) {
                longest = evenPalindrome;
            }
        }

        return longest;
    }

    public static void main(String[] args) {
        String input = "babad";
        System.out.println("A maior substring palindrômica é: " + longestPalindromicSubstring(input));
    }
}

File: ChatGPT/Intermediárias/longest_palindromic_substring/longest_palindromic_substring.cpp
#include <iostream>
#include <string>

using namespace std;

// Função para expandir ao redor do centro e encontrar o palíndromo
string expandAroundCenter(const string& s, int left, int right) {
    while (left >= 0 && right < s.length() && s[left] == s[right]) {
        left--;
        right++;
    }
    return s.substr(left + 1, right - left - 1);
}

// Função para encontrar a maior substring palindrômica
string longestPalindromicSubstring(const string& s) {
    if (s.length() < 2) {
        return s;
    }

    string longest;
    for (int i = 0; i < s.length(); i++) {
        // Encontrar o maior palíndromo com um centro único
        string oddPalindrome = expandAroundCenter(s, i, i);
        if (oddPalindrome.length() > longest.length()) {
            longest = oddPalindrome;
        }

        // Encontrar o maior palíndromo com um centro duplo
        string evenPalindrome = expandAroundCenter(s, i, i + 1);
        if (evenPalindrome.length() > longest.length()) {
            longest = evenPalindrome;
        }
    }

    return longest;
}

// Função principal para testar o algoritmo de substring palindrômica
int main() {
    string input = "babad";
    cout << "A maior substring palindrômica é: " << longestPalindromicSubstring(input) << endl;
    return 0;
}
